source/common/http/conn_pool_grid.cc

```diff
@@ -54,8 +56,12 @@ void ConnectivityGrid::WrapperCallbacks::ConnectionAttemptCallbacks::onPoolFailu
 void ConnectivityGrid::WrapperCallbacks::onConnectionAttemptFailed(
     ConnectionAttemptCallbacks* attempt, ConnectionPool::PoolFailureReason reason,
     absl::string_view transport_failure_reason, Upstream::HostDescriptionConstSharedPtr host) {
+  ASSERT(host == grid_.host_);
   ENVOY_LOG(trace, "{} pool failed to create connection to host '{}'.",
             describePool(attempt->pool()), grid_.host_->hostname());
+  if (&(attempt->pool()) == &http3_pool_) {
```

Comment:
it's a bit hard to tell because things are mocked so much, but I think we can't actually hit this case until we remove the attempt->cancellable_->cancel() from onPoolReady.

Basically right now if H2 connects before H3, we cancel the H3 and it will never actually finish.
so I think we need to amend that cancel call with "if this is H2 connecting, tell the H3 wrapper callback to not do upcalls on completion"

Reply:
I think we hit this codepath in the case where the HTTP/3 job fails first, and then we start the TCP job. In particular, this happens in ConnectivityGridTest.FailureThenSuccessSerial (added an ASSERT(false) in this code and it fired).

That being said, the case where H2 succeeds before H3 fails is not yet handled.